Radiation pulses will yield to MC for every turf they iterate through.

Pickle-Coding opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Changes radiation pulses so that they store the turfs they need to irradiate, iterate through them and remove them from the list of turfs to irradiate, then yield to MC if MC got angry, and put the radiation pulse with the remaining turfs to irradiate to the back of the radiation processing list.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
So radiation pulses will listen to the MC more on when to stop.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
code: Radiation pulses will ask MC if they should continue processing for every turf they iterate through, rather than just every pulse.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Pickle-Coding committed:
Be less scary.

Changes radiation pulses so that they store the turfs they need to irradiate, iterate through them and remove them from the list of turfs to irradiate, then yield to MC if MC got angry, and put the radiation pulse with the remaining turfs to irradiate to the back of the radiation processing list.

Pickle-Coding committed:
hhhhhhhhh

Removes unnecessary "var/" in /proc/get_rad_turfs() for the max_range parameter.

Pickle-Coding commented:
This might actually make a moving object more likely to get irradiated if they move to another turf while the pulse hasn't been complete yet. Should probably do somethint about that.

Pickle-Coding commented:
Actually I am not sure if there is a solution that is performant and doesn't have other issues, so I'll just leave it to maintainer discretion on whether I should store and iterate through movables instead of turfs or not.

LemonInTheDark commented on code/__HELPERS/radiation.dm:
```diff
@@ -35,6 +35,7 @@
 	pulse_information.threshold = threshold
 	pulse_information.chance = chance
 	pulse_information.minimum_exposure_time = minimum_exposure_time
+	pulse_information.turfs_to_process = get_rad_turfs(source, max_range)
```
I'd recommend `RANGE_TURFS` here, which uses the internal block() proc, as laid out in the 8 year old profile series done by Rastaf0 <https://gist.github.com/Giacom/be635398926bb463b42a>
It's a bit faster, and also cleaner I think

LemonInTheDark commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -28,65 +28,69 @@ SUBSYSTEM_DEF(radiation)
 
 /datum/controller/subsystem/radiation/proc/pulse(atom/source, datum/radiation_pulse_information/pulse_information)
 	var/list/cached_rad_insulations = list()
+	while(pulse_information.turfs_to_process.len)
+		var/turf/turf_to_irradiate = popleft(pulse_information.turfs_to_process)
```
if you're gonna iterate turfs_to_process, you should cache it into a local list variable. I'm not joking it's an unnerving amount faster.
Also rather then using popleft, you could count how many things we've iterated, then cut the list right at the end, either when we're done processing or when we yield. S cleaner and a bit cheaper that way.
Also no proc overhead

LemonInTheDark commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -28,65 +28,69 @@ SUBSYSTEM_DEF(radiation)
+			if (irradiate_after_basic_checks(target))
+				target.investigate_log("was irradiated by [source].", INVESTIGATE_RADIATION)
+		if(MC_TICK_CHECK)
+			processing += pulse_information
+			return
```
Rather then reinserting pulse information back into the processing list, why not only remove it from the processing list if pulse() doesn't lead to MC_TICK_CHECK being true? Little cleaner that way

LemonInTheDark reviewed (changes requested):
Good work, some minor comments

Mothblocks commented:
> This might actually make a moving object more likely to get irradiated if they move to another turf while the pulse hasn't been complete yet

This is the primary thing I told Lemon to ensure would not happen FYI--if a player moves out of the way of a radiation source, they should not become irradiated a decent time after because of CPU limitations.

Pickle-Coding commented:
> > This might actually make a moving object more likely to get irradiated if they move to another turf while the pulse hasn't been complete yet
> 
> 
> 
> This is the primary thing I told Lemon to ensure would not happen FYI--if a player moves out of the way of a radiation source, they should not become irradiated a decent time after because of CPU limitations.

I don't think it's likely for someone to get irradiated after they move outside of the radiation source, as it checks for things inside the turfs. The problem with this one is the chance of getting irradiated can be higher than it should be if you're inside the range, if you move into an unprocessed turf while getting irradiated.

LemonInTheDark commented:
> This is the primary thing I told Lemon to ensure would not happen FYI--if a player moves out of the way of a radiation source, they should not become irradiated a decent time after because of CPU limitations.

The issue here is essentially the inverse, with players moving from a processed turf to an unprocessed turf, or an unprocessed turf to a processed one.
That or moves INTO the path of the wave.

You won't be dosed if you've left the area tho

Mothblocks commented:
That's a bit crummy but not as bad I think

Pickle-Coding committed:
Reviews and stuff.

Radiation pulses pulse information will no longer popleft the processing list. It will delete itself once there are no turfs to process, or if the source is null.

Caches the turfs to process to loop through, adds a counter for how many turfs it iterated through, and cuts the turfs to process by the turfs that got iterated.

Pickle-Coding committed:
I forgot.

I had a skill issue with lists, and thought MC tick check kept becoming true before it could get cut, but it was just my skill issue of using Cut() thinking it would return something.

Removes the unnecessary turf to process length check, and just removes the first item of the processing list if MC tick check is false.

LemonInTheDark commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -10,83 +10,95 @@ SUBSYSTEM_DEF(radiation)
 
 /datum/controller/subsystem/radiation/proc/pulse(atom/source, datum/radiation_pulse_information/pulse_information)
 	var/list/cached_rad_insulations = list()
+	var/list/cached_turfs_to_process = pulse_information.turfs_to_process.Copy()
```
why copy this?

Pickle-Coding replied:
To cache it into a list.

LemonInTheDark replied:
It's dumber then that. Just setting a proc var to a datum var caches it and makes access times faster

LemonInTheDark commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -10,83 +10,95 @@ SUBSYSTEM_DEF(radiation)
+				target.investigate_log("was irradiated by [source].", INVESTIGATE_RADIATION)
+
+		if(MC_TICK_CHECK)
+			break
```
make sure to cut here too, since if we yield back we want to remove the turfs we processed

Pickle-Coding replied:
Doesn't break just break out of the loop, then the cut gets called?

LemonInTheDark replied:
Oh true true didn't notice that thought it was a return

LemonInTheDark commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -10,83 +10,95 @@ SUBSYSTEM_DEF(radiation)
+		if(MC_TICK_CHECK)
+			break
+
+	pulse_information.turfs_to_process.Cut(1, turfs_iterated + 1)
```
why not use the cached thing here?

Pickle-Coding replied:
What do you mean by using it here?

LemonInTheDark replied:
See 2 above

Pickle-Coding committed:
Don't copy.

We don't need to copy turfs_to_process as it won't change during the loop. Just need to cache it for accessing.

Pickle-Coding committed:
Cool.

Now that cached_turfs_to_process is a reference to pulse_information.turfs_to_process, we can just call cut on that instead of the datum variable.

Kylerace commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -10,83 +10,95 @@ SUBSYSTEM_DEF(radiation)
+	var/list/cached_turfs_to_process = pulse_information.turfs_to_process
+	var/turfs_iterated = 0
+	for (var/turf/turf in cached_turfs_to_process)
+		var/turf/turf_to_irradiate = turf
```
guh? whats the point of this just do as anything

Pickle-Coding replied:
I don't remember why I added that, but it doesn't serve a purpose so I removed that variable.

Pickle-Coding committed:
Removes useless variable.

For some reason I added a variable that just gets assigned to the for loop variable, which is kinda pointless. Removes that variable and renames the for loop variable to that variable.

LemonInTheDark commented on code/controllers/subsystem/radiation.dm:
```diff
@@ -10,83 +10,94 @@ SUBSYSTEM_DEF(radiation)
 	var/list/cached_rad_insulations = list()
+	var/list/cached_turfs_to_process = pulse_information.turfs_to_process
+	var/turfs_iterated = 0
+	for (var/turf/turf_to_irradiate as anything in cached_turfs_to_process)
+		turfs_iterated += 1
```
It is worth noting that iterations in this form make a copy of the list, so they are not preferable if iterating and yielding over a LARGE list
This is not that tho, by large I mean like, 20k size. See the garbage subsystem.
That is why you'll see the while (thing.length) thing[thing.length] pattern so much though
